components/scream/src/share/util/scream_common_physics_impl.hpp

```diff
@@ -0,0 +1,263 @@
+template<typename DeviceT>
+template<typename ScalarT>
+KOKKOS_FUNCTION
+ScalarT PhysicsFunctions<DeviceT>::calculate_theta_from_T(const ScalarT& T_mid, const ScalarT& p_mid)
+{
+  ScalarT exner = exner_function(p_mid);
+  ScalarT theta = T_mid/exner;
+  // Return theta
+  return theta;
+}
+// For operations on a full column at a time:
+template<typename DeviceT>
+template<typename ScalarT, typename InputProviderT, typename InputProviderP>
+KOKKOS_FUNCTION
+void PhysicsFunctions<DeviceT>::calculate_theta_from_T(const MemberType& team,
+                                                 const InputProviderT& T_mid,
+                                                 const InputProviderP& p_mid,
+                                                 const view_1d<ScalarT>& theta)
+{
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,theta.extent(0)),
+                       [&] (const int k) {
+    theta(k) = calculate_theta_from_T(T_mid(k),p_mid(k)); 
+  });
+}
+
+//-----------------------------------------------------------------------------------------------//
+// Converts potential temperature into temperature
+// The result is the temperature, units in [K]
+// The inputs are
+//   theta is the potential temperature, units in [K]
+//   p_mid is the atmospheric pressure, units in [Pa].  p_mid is used in Exners function using `exner` defined above.
+template<typename DeviceT>
+template<typename ScalarT>
+KOKKOS_FUNCTION
+ScalarT PhysicsFunctions<DeviceT>::calculate_T_from_theta(const ScalarT& theta, const ScalarT& p_mid)
+{
+  ScalarT exner = exner_function(p_mid);
+  ScalarT T_mid = theta*exner;
+  // Return T
+  return T_mid;
+}
+// For operations on a full column at a time:
+template<typename DeviceT>
+template<typename ScalarT, typename InputProviderT, typename InputProviderP>
+KOKKOS_FUNCTION
+void PhysicsFunctions<DeviceT>::calculate_T_from_theta(const MemberType& team,
+                                                 const InputProviderT& theta,
+                                                 const InputProviderP& p_mid,
+                                                 const view_1d<ScalarT>& T_mid)
+{
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,T_mid.extent(0)),
+                       [&] (const int k) {
+    T_mid(k) = calculate_T_from_theta(theta(k),p_mid(k)); 
+  });
+}
+//-----------------------------------------------------------------------------------------------//
+// Compute temperature from virtual temperature
+// The result unit is in [K]
+// The inputs are
+//   T_virtual is the virtual temperature.  Units in [K].
+//   qv        is the water vapor mass mixing ratio.  Units in [kg/kg]
+template<typename DeviceT>
+template<typename ScalarT>
+KOKKOS_FUNCTION
+ScalarT PhysicsFunctions<DeviceT>::calculate_temperature_from_virtual_temperature(const ScalarT& T_virtual, const ScalarT& qv)
+{
+  using C = scream::physics::Constants<Real>;
+  static const ScalarT ep_2(C::ep_2);
```

Comment:
I originally had `ScalarT ep_2 = C:ep_2` but it doesn't work when ScalarT is a pack, since C::ep_2 is not of the same type.  This way, C::ep_2 is the constructor for ep_2 irregardless if ScalarT is a Real or a pack.

Reply:
per `ep_2`: I believe this is because the variable is generally written as the greek letter epsilon, but that is already in use so this is a convenient shorthand.  In this case I am taking it to match the name in physics_constants.hpp.  I'm open to changing the name locally, or if there is agreement in the group changing it in physics_constants.hpp as well.

Reply:
`static constexpr ScalarT ep_2 = C::ep_2;` this actually breaks in the build with an error 
```
error: expression must have a constant value
```